presto-parser/src/main/java/com/facebook/presto/sql/parser/ErrorHandler.java

```diff
@@ -0,0 +1,207 @@
+        extends BaseErrorListener
+{
+    private final Map<Integer, String> specialRules;
+    private final Map<Integer, String> specialTokens;
+    private final Set<Integer> ignoredRules;
+
+    public ErrorHandler(Map<Integer, String> specialRules, Map<Integer, String> specialTokens, Set<Integer> ignoredRules)
+    {
+        this.specialRules = new HashMap<>(specialRules);
+        this.specialTokens = specialTokens;
+        this.ignoredRules = new HashSet<>(ignoredRules);
+    }
+
+    @Override
+    public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int line, int charPositionInLine, String message, RecognitionException e)
+    {
+        Parser parser = (Parser) recognizer;
+
+        ATN atn = parser.getATN();
+        ATNState currentState = atn.states.get(parser.getState());
+
+        Analyzer analyzer = new Analyzer(parser.getTokenStream(), atn, parser.getVocabulary(), specialRules, specialTokens, ignoredRules);
+
+        Map<Integer, Set<String>> candidates = new HashMap<>();
+        analyzer.process(candidates, currentState, parser.getCurrentToken().getTokenIndex(), parser.getContext());
+
+        // pick the candidate tokens associated largest token index processed (i.e., the path that consumed the most input)
+        String expected = candidates.entrySet().stream()
+                .max(Comparator.comparing(Map.Entry::getKey))
+                .get()
+                .getValue().stream()
+                .sorted()
+                .collect(Collectors.joining(", "));
+
+        throw new ParsingException(
+                String.format("mismatched input '%s'. Expecting: %s", ((Token) offendingSymbol).getText(), expected),
+                e,
+                line,
+                charPositionInLine);
+    }
+
+    private static class ParsingState
+    {
+        public final ATNState state;
+        public final int tokenIndex;
+
+        public ParsingState(ATNState state, int tokenIndex)
+        {
+            this.state = state;
+            this.tokenIndex = tokenIndex;
+        }
+    }
+
+    private static class Analyzer
+    {
+        private final TokenStream stream;
+        private final ATN atn;
+        private final Vocabulary vocabulary;
+        private final Map<Integer, String> specialRules;
+        private final Map<Integer, String> specialTokens;
+        private final Set<Integer> ignoredRules;
+
+        public Analyzer(TokenStream stream, ATN atn, Vocabulary vocabulary, Map<Integer, String> specialRules, Map<Integer, String> specialTokens, Set<Integer> ignoredRules)
+        {
+            this.stream = stream;
+            this.atn = atn;
+            this.vocabulary = vocabulary;
+            this.specialRules = specialRules;
+            this.specialTokens = specialTokens;
+            this.ignoredRules = ignoredRules;
+        }
+
+        private Set<Integer> process(Map<Integer, Set<String>> candidates, ATNState startState, int tokenIndex, RuleContext next)
```

Comment:
Maybe I'm missing something subtle, but why not just have `candidates` be a field on this class?

Reply:
It could... legacy from when I was prototyping it.